### PR TITLE
Make m_mempool optional in CChainState

### PR DESCRIPTION
**Pull request by jamesob**
Make `CChainState::m_mempool` optional by making it a pointer instead of a reference. This will allow a simplification to assumeutxo semantics (see https://github.com/bitcoin/bitcoin/pull/15606#pullrequestreview-692965905) and help facilitate the `-nomempool` option.

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -2565,12 +2570,12 @@ bool CChainState::ActivateBestChainStep(BlockValidationState& state, CBlockIndex
         }
     }
 
-    if (fBlocksDisconnected) {
+    if (fBlocksDisconnected && m_mempool) {
         // If any blocks were disconnected, disconnectpool may be non empty.  Add
         // any disconnected transactions back to the mempool.
-        UpdateMempoolForReorg(*this, m_mempool, disconnectpool, true);
+        mempoolHandleReorg(disconnectpool, false);
```
Comment (MarcoFalke) [inline src/validation.cpp:2576]: why the change from `true` -> `false`?

File: src/validation.cpp [hunk 2]
```diff
@@ -2642,7 +2647,8 @@ bool CChainState::ActivateBestChain(BlockValidationState& state, std::shared_ptr
 
         {
             LOCK(cs_main);
-            LOCK(m_mempool.cs); // Lock transaction pool for at least as long as it takes for connectTrace to be consumed
+            // Lock transaction pool for at least as long as it takes for connectTrace to be consumed
+            if (m_mempool) LOCK(m_mempool->cs);
```
Comment (MarcoFalke) [inline src/validation.cpp:2651]: Nice try, but leaving out the `{}` won't prevent the lock from being closed in the same line.

File: src/validation.cpp [hunk 3]
```diff
@@ -2792,7 +2803,9 @@ bool CChainState::InvalidateBlock(BlockValidationState& state, CBlockIndex* pind
         LimitValidationInterfaceQueue();
 
         LOCK(cs_main);
-        LOCK(m_mempool.cs); // Lock for as long as disconnectpool is in scope to make sure UpdateMempoolForReorg is called after DisconnectTip without unlocking in between
+        // Lock for as long as disconnectpool is in scope to make sure UpdateMempoolForReorg is
+        // called after DisconnectTip without unlocking in between
+        if (m_mempool) LOCK(m_mempool->cs);
```
Comment (MarcoFalke) [inline src/validation.cpp:2808]: same

File: src/validation.cpp [hunk 4]
```diff
@@ -4492,6 +4506,41 @@ bool CChainState::ResizeCoinsCaches(size_t coinstip_size, size_t coinsdb_size)
     return ret;
 }
 
+void CChainState::mempoolHandleReorg(DisconnectedBlockTransactions& disconnectpool, bool add_to_mempool) {
```
Comment (MarcoFalke) [inline src/validation.cpp:4509]:
```suggestion
void CChainState::mempoolHandleReorg(DisconnectedBlockTransactions& disconnectpool, bool add_to_mempool)
{
```

properly format new code?

File: src/validation.h [hunk 5]
```diff
@@ -600,7 +601,7 @@ class CChainState
     //! CChainState instances.
     BlockManager& m_blockman;
 
-    explicit CChainState(CTxMemPool& mempool, BlockManager& blockman, std::optional<uint256> from_snapshot_blockhash = std::nullopt);
+    explicit CChainState(CTxMemPool* mempool, BlockManager& blockman, std::optional<uint256> from_snapshot_blockhash = std::nullopt);
```
Comment (MarcoFalke) [inline src/validation.h:604]: Would it be possible to combine the args so that *either* a mempool, or a hash can be passed, but not both? I think std::variant can do this

**Review by maflcko — COMMENTED** [review]
Concept ACK

**Comment by ryanofsky** [comment]
Concept ACK. I think you can drop the second commit and add a `LOCK_RETURNED` helper method to be able to keep using lock annotations when m_mempool is null:

```c++
RecursiveMutex* MempoolMutex() const LOCK_RETURNED(m_mempool->cs) { return m_mempool ? &m_mempool->cs : nullptr; }
```

Seems to work for me in commit 8db5953e2d666fb9c25ae8f0df5b4c5c5897cd58

File: src/validation.cpp (hunk 1 above)
Comment (jamesob) [reply src/validation.cpp:2576]: Hey somebody's gotta keep review fun!

File: src/validation.cpp (hunk 2 above)
Comment (jamesob) [reply src/validation.cpp:2651]: Oh yeah d'oh. Sure could've used those missing annotations there.

**Comment by jamesob** [comment]
Rebased to remove the second commit per Russ' suggestion, which also addresses most of Marco's feedback.

> Seems to work for me in commit 8db5953

Wow, that was wizardly. Thanks @ryanofsky!

> I think std::variant can do this

I'll look into this tomorrow.

File: src/validation.h [hunk 6]
```diff
@@ -907,7 +913,7 @@ class ChainstateManager
     //                                  constructor
     //! @param[in] snapshot_blockhash   If given, signify that this chainstate
     //!                                 is based on a snapshot.
-    CChainState& InitializeChainstate(CTxMemPool& mempool, const std::optional<uint256>& snapshot_blockhash = std::nullopt)
+    CChainState& InitializeChainstate(CTxMemPool* mempool, const std::optional<uint256>& snapshot_blockhash = std::nullopt)
```
Comment (naumenkogs) [inline src/validation.h:916]: Similarly to [this comment](https://github.com/bitcoin/bitcoin/pull/22415/commits/b17bb4c37d58ca49e877983a632bfcaa493be993#r665701402), can we somehow reflect that if mempool is nullptr, the second argument doesn't make sense?

**Comment by naumenkogs** [comment]
Light code review ACK b17bb4c37d58ca49e877983a632bfcaa493be993

File: src/validation.h [hunk 7]
```diff
@@ -798,6 +799,11 @@ class CChainState
 
     bool LoadBlockIndexDB() EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
+    //! Indirection necessary to make lock annotations work with an optional mempool.
+    RecursiveMutex* MempoolMutex() const LOCK_RETURNED(m_mempool->cs) {
```
Comment (MarcoFalke) [inline src/validation.h:803]:
```suggestion
    RecursiveMutex* MempoolMutex() const LOCK_RETURNED(m_mempool->cs)
    {
```

clang-format?

**Review by maflcko — COMMENTED** [review]
Nice solution with the `LOCK_RETURNED`

**Comment by jonatack** [comment]
Concept ACK, will review on next push.

**Review by ryanofsky — APPROVED** [review]
Code review ACK b17bb4c37d58ca49e877983a632bfcaa493be993

**Comment by jamesob** [comment]
Rebased.

> Note: Naively I'd expect the mempool pointer for the activated snapshot to be non-null, but I guess it could be set later when the snapshot is fully initialized.

This is a good point. I'd come up with a commit that implements the `std::variant` feedback (and am glad to have for learning about that), but @ryanofsky's point here makes that change inappropriate: with assumeutxo, we may actually want the snapshot chain to have a non-null mempool if we're in, say, init.cpp and loading a snapshot chainstate that is active but not yet at tip.

I've covered all of the other feedback aside from the lock assertion for the reasons Russ mentioned; maybe we can get around to it later.

File: src/validation.h (hunk 5 above)
Comment (jamesob) [reply src/validation.h:604]: Not going this route for reasons mentioned in a recent comment: passing both is a valid approach if we are initializing a not-yet-to-tip snapshot chainstate in init.cpp.

File: src/init.cpp [hunk 8]
```diff
@@ -1349,7 +1349,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
             const int64_t load_block_index_start_time = GetTimeMillis();
             try {
                 LOCK(cs_main);
-                chainman.InitializeChainstate(*Assert(node.mempool));
+                chainman.InitializeChainstate(Assert(node.mempool).get());
```
[the remark this answers is not in the document]
Comment (jamesob) [reply src/init.cpp]: Fixed, thanks.

**Review by ryanofsky — APPROVED** [review]
Code review ACK 8511429f3bc480ddeef06c23810b3c47fa8513ee. Just some minor tweaks since last review: whitespace, unique_ptr assert, null mempool args

**Review by jonatack — COMMENTED** [review]
ACK 8511429f3bc480ddeef06c23810b3c47fa8513ee code review, debug build, running bitcoind on signet as a sanity check, operation nominal

File: src/validation.h (hunk 6 above)
Comment (naumenkogs) [reply src/validation.h:917]: What do you think about making this std::optional<CTxMemPool, std::optional<uint256>>?

In this case, snapshot_blockhash can't be possibly specified without mempool, and that's what I originally suggested.

File: src/validation.h [hunk 9]
```diff
@@ -907,7 +914,7 @@ class ChainstateManager
     //                                  constructor
     //! @param[in] snapshot_blockhash   If given, signify that this chainstate
     //!                                 is based on a snapshot.
-    CChainState& InitializeChainstate(CTxMemPool& mempool, const std::optional<uint256>& snapshot_blockhash = std::nullopt)
+    CChainState& InitializeChainstate(CTxMemPool* mempool, const std::optional<uint256>& snapshot_blockhash = std::nullopt)
```
Comment (naumenkogs) [inline src/validation.h:917]: What do you think about making this std::optional<CTxMemPool, std::optional>?

In this case, snapshot_blockhash can't be possibly specified without mempool, and that's what I originally suggested.

File: src/validation.cpp [hunk 10]
```diff
@@ -2288,7 +2291,7 @@ bool CChainState::DisconnectTip(BlockValidationState& state, DisconnectedBlockTr
         while (disconnectpool->DynamicMemoryUsage() > MAX_DISCONNECTED_TX_POOL_SIZE * 1000) {
             // Drop the earliest entry, and remove its children from the mempool.
             auto it = disconnectpool->queuedTx.get<insertion_order>().begin();
-            m_mempool.removeRecursive(**it, MemPoolRemovalReason::REORG);
+            if (m_mempool) m_mempool->removeRecursive(**it, MemPoolRemovalReason::REORG);
```
Comment (jnewbery) [inline src/validation.cpp:2294]: This `if (m_mempool)` condition can be at the level of the `if (disconnectpool)` above:

```diff
         return false;
     }
 
-    if (disconnectpool) {
+    if (m_mempool && disconnectpool) {
         // Save transactions to re-add to mempool at end of reorg
         for (auto it = block.vtx.rbegin(); it != block.vtx.rend(); ++it) {
             disconnectpool->addTransaction(*it);
@@ -2291,7 +2291,7 @@ bool CChainState::DisconnectTip(BlockValidationState& state, DisconnectedBlockTr
         while (disconnectpool->DynamicMemoryUsage() > MAX_DISCONNECTED_TX_POOL_SIZE * 1000) {
             // Drop the earliest entry, and remove its children from the mempool.
             auto it = disconnectpool->queuedTx.get<insertion_order>().begin();
-            if (m_mempool) m_mempool->removeRecursive(**it, MemPoolRemovalReason::REORG);
+            m_mempool->removeRecursive(**it, MemPoolRemovalReason::REORG);
             disconnectpool->removeEntry(it);
         }
     }
```

That would potentially save some work updating the `disconnectpool` data structure, which will never get used.

Even better (but not necessarily part of this PR) would be to only create a `DisconnectedBlockTransactions` if `m_mempool` is non-null, and then update ` UpdateMempoolForReorg()` and ` DisconnectedBlockTransactions()` to take optional `DisconnectedBlockTransactions*`. There's no need to create and maintain a `DisconnectedBlockTransactions` if we're never going to use it.

File: src/validation.cpp [hunk 11]
```diff
@@ -2205,11 +2206,13 @@ static void AppendWarning(bilingual_str& res, const bilingual_str& warn)
 }
 
 /** Check warning conditions and do some notifications on new chain tip set. */
-static void UpdateTip(CTxMemPool& mempool, const CBlockIndex* pindexNew, const CChainParams& chainParams, CChainState& active_chainstate)
+static void UpdateTip(CTxMemPool* mempool, const CBlockIndex* pindexNew, const CChainParams& chainParams, CChainState& active_chainstate)
```
Comment (jnewbery) [inline src/validation.cpp:2209]: This could be a member of `CChainState` to avoid passing `mempool`, `chainParams` and `active_chainstate`.

File: src/validation.cpp [hunk 12]
```diff
@@ -2401,7 +2404,7 @@ bool CChainState::ConnectTip(BlockValidationState& state, CBlockIndex* pindexNew
     int64_t nTime5 = GetTimeMicros(); nTimeChainState += nTime5 - nTime4;
     LogPrint(BCLog::BENCH, "  - Writing chainstate: %.2fms [%.2fs (%.2fms/blk)]\n", (nTime5 - nTime4) * MILLI, nTimeChainState * MICRO, nTimeChainState * MILLI / nBlocksTotal);
     // Remove conflicting transactions from the mempool.;
-    m_mempool.removeForBlock(blockConnecting.vtx, pindexNew->nHeight);
+    if (m_mempool) m_mempool->removeForBlock(blockConnecting.vtx, pindexNew->nHeight);
     disconnectpool.removeForBlock(blockConnecting.vtx);
```
Comment (jnewbery) [inline src/validation.cpp:2408]: No need to update `disconnectpool` if the mempool doesn't exist:

```suggestion
    if (m_mempool) {
        m_mempool->removeForBlock(blockConnecting.vtx, pindexNew->nHeight);
        disconnectpool.removeForBlock(blockConnecting.vtx);
    }
```

File: src/validation.cpp [hunk 13]
```diff
@@ -2507,7 +2510,7 @@ bool CChainState::ActivateBestChainStep(BlockValidationState& state, CBlockIndex
         if (!DisconnectTip(state, &disconnectpool)) {
             // This is likely a fatal error, but keep the mempool consistent,
             // just in case. Only remove from the mempool in this case.
-            UpdateMempoolForReorg(*this, m_mempool, disconnectpool, false);
+            if (m_mempool) UpdateMempoolForReorg(*this, *m_mempool, disconnectpool, false);
```
Comment (jnewbery) [inline src/validation.cpp:2513]: What do you think about updating `UpdateMempoolForReorg` to either be a member of `CChainState` or take `CTxMemPool` as an optional pointer, and then return immediately if mempool is null. That would avoid scattering so many `if (m_mempool)` checks across multiple validation functions.

**Review by jnewbery — COMMENTED** [review]
Concept ACK. A few non-blocking suggestions inline.

File: src/validation.cpp [hunk 14]
```diff
@@ -2053,7 +2054,7 @@ bool CChainState::FlushStateToDisk(
         bool fFlushForPrune = false;
         bool fDoFullFlush = false;
 
-        CoinsCacheSizeState cache_state = GetCoinsCacheSizeState(&m_mempool);
+        CoinsCacheSizeState cache_state = GetCoinsCacheSizeState(m_mempool);
```
[the remark this answers is not in the document]
Comment (jamesob) [reply src/validation.cpp]: Good call! Done.

File: src/validation.cpp (hunk 10 above)
Comment (jamesob) [reply src/validation.cpp:2294]: Yup, also a good call. Fixed. Initially wasn't sure if there was some other purpose we might be using the disconnectpool for (since it's sometimes passed in as a mutable parameter) but after looking around it's clearly just used for mempool maintenance.

File: src/validation.cpp (hunk 11 above)
Comment (jamesob) [reply src/validation.cpp]: Agreed, done.

File: src/validation.cpp (hunk 12 above)
Comment (jamesob) [reply src/validation.cpp]: :+1:

File: src/validation.cpp (hunk 13 above)
Comment (jamesob) [reply src/validation.cpp]: I went ahead and threw it onto `CChainState` since I think that makes sense; can strip off the commit if there are any objections.

File: src/validation.h (hunk 9 above)
Comment (jamesob) [reply src/validation.h:924]: There are cases when we want to call this with a snapshot blockhash but not a mempool - during snapshot load time when we need a chainstate instance to populate/validate but don't want to yet make it the active chainstate in case loading the UTXO snapshot fails.

**Comment by jamesob** [comment]
Thanks for the prompt attention, good people. I've pushed a rebase that includes pretty much all of @jnewbery's feedback. This widens the scope of the PR to encompass a bit of refactoring, but I think it's all reasonable stuff to include here.

File: src/validation.cpp [hunk 15]
```diff
@@ -342,11 +342,15 @@ static bool IsCurrentForFeeEstimation(CChainState& active_chainstate) EXCLUSIVE_
  * Passing fAddToMempool=false will skip trying to add the transactions back,
  * and instead just erase from the mempool as needed.
  */
-
-static void UpdateMempoolForReorg(CChainState& active_chainstate, CTxMemPool& mempool, DisconnectedBlockTransactions& disconnectpool, bool fAddToMempool) EXCLUSIVE_LOCKS_REQUIRED(cs_main, mempool.cs)
+void CChainState::UpdateMempoolForReorg(
+    CChainState& active_chainstate,
```
Comment (ryanofsky) [inline src/validation.cpp:346]: In commit "refactor: move UpdateMempoolForReorg into CChainState" (7cdcd08d665570167b06a93e003369f7fa833cec)

`active_chainstate` is always `*this` right? Can the parameter be dropped?

**Review by ryanofsky — COMMENTED** [review]
Code review ACK 29acd850fb0539b8358a451089ef7b1c583dd1a1. Simplifications from John make the change nicer.

File: src/validation.cpp (hunk 15 above)
Comment (jamesob) [reply src/validation.cpp]: Good catch, thanks. Always weary to assume active_chainstate == *this, but in this case (as you probably realized) only the active ever has a mempool, so it's a fine substitution.

**Comment by jamesob** [comment]
Rebase to incorporate minor changes from @ryanofsky's feedback (https://github.com/bitcoin/bitcoin/pull/22415#discussion_r667134941).

File: src/validation.cpp [hunk 16]
```diff
@@ -2280,15 +2283,15 @@ bool CChainState::DisconnectTip(BlockValidationState& state, DisconnectedBlockTr
...
+    if (disconnectpool && m_mempool) {
         // Save transactions to re-add to mempool at end of reorg
         for (auto it = block.vtx.rbegin(); it != block.vtx.rend(); ++it) {
             disconnectpool->addTransaction(*it);
         }
         while (disconnectpool->DynamicMemoryUsage() > MAX_DISCONNECTED_TX_POOL_SIZE * 1000) {
             // Drop the earliest entry, and remove its children from the mempool.
             auto it = disconnectpool->queuedTx.get<insertion_order>().begin();
-            m_mempool.removeRecursive(**it, MemPoolRemovalReason::REORG);
+            if (m_mempool) m_mempool->removeRecursive(**it, MemPoolRemovalReason::REORG);
```
Comment (naumenkogs) [inline src/validation.cpp:2299]: This new if here is not needed, it's augmented by the check you added above (`if (disconnectpool && m_mempool) {`)?

File: src/validation.cpp [hunk 17]
```diff
@@ -342,11 +342,14 @@ static bool IsCurrentForFeeEstimation(CChainState& active_chainstate) EXCLUSIVE_
  * Passing fAddToMempool=false will skip trying to add the transactions back,
  * and instead just erase from the mempool as needed.
  */
-
-static void UpdateMempoolForReorg(CChainState& active_chainstate, CTxMemPool& mempool, DisconnectedBlockTransactions& disconnectpool, bool fAddToMempool) EXCLUSIVE_LOCKS_REQUIRED(cs_main, mempool.cs)
+void CChainState::UpdateMempoolForReorg(
```
Comment (naumenkogs) [inline src/validation.cpp:345]: Mixed feelings about this change. I think a reader might get confused: "`UpdateMempoolForReorg`? But I don't have mempool."

I would slightly prefer changing the name to `UpdateMempoolForReorgIfExists` or `TryUpdateMempoolForReorg`, but feel free to ignore.

File: src/validation.h [hunk 18]
```diff
@@ -587,8 +587,9 @@ class CChainState
      */
     mutable std::atomic<bool> m_cached_finished_ibd{false};
 
-    //! mempool that is kept in sync with the chain
-    CTxMemPool& m_mempool;
+    //! Optional mempool that is kept in sync with the chain.
+    //! Only the active chainstate has a mempool.
+    CTxMemPool* m_mempool;
```
Comment (jnewbery) [inline src/validation.h:592]: This could be a const pointer:

```suggestion
    CTxMemPool* const m_mempool;
```

since it's set in the initializer list and never modified.

File: src/validation.cpp [hunk 19]
```diff
@@ -4763,7 +4772,7 @@ bool ChainstateManager::ActivateSnapshot(
     }
 
     auto snapshot_chainstate = WITH_LOCK(::cs_main, return std::make_unique<CChainState>(
-            this->ActiveChainstate().m_mempool, m_blockman, base_blockhash));
+            nullptr, m_blockman, base_blockhash));
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/validation.cpp]: I was also surprised by this. Is the snapshot chainstate the one that eventually becomes our tip? In that case, don't we want it to have a mempool?

File: src/validation.cpp [hunk 20]
```diff
@@ -329,7 +329,8 @@ static bool IsCurrentForFeeEstimation(CChainState& active_chainstate) EXCLUSIVE_
     return true;
 }
 
-/* Make mempool consistent after a reorg, by re-adding or recursively erasing
+/**
```
Comment (jnewbery) [inline src/validation.cpp:332]: nit: doxygen function comments should go with the function declaration. Do you mind moving this to the header file now that the declaration is there?

File: src/validation.h [hunk 21]
```diff
@@ -798,6 +797,20 @@ class CChainState
...
+    {
+        return m_mempool ? &m_mempool->cs : nullptr;
+    }
+
+    void UpdateMempoolForReorg(
+        DisconnectedBlockTransactions& disconnectpool,
+        bool fAddToMempool) EXCLUSIVE_LOCKS_REQUIRED(cs_main, m_mempool->cs);
+
+    /** Check warning conditions and do some notifications on new chain tip set. */
+    void UpdateTip(const CBlockIndex* pindexNew, CChainState& active_chainstate)
```
Comment (jnewbery) [inline src/validation.h:811]: No need for the `CChainState& active_chainstate` argument - this is a private method in `CChainState` where the `active_chainstate` arg is always `*this`. You can remove the argument and use the implicit `this` inside the function.

**Review by jnewbery — COMMENTED** [review]
Looking good. Just a few small comments inline.

File: src/validation.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (jamesob) [reply src/validation.cpp]: Yeah, this is a little confusing absent context, but basically the idea per @ryanofsky's feedback is that mempool presence should be used to indicate chainstate activeness, so we don't want to give the snapshot chainstate a mempool until we consider it active.

File: src/validation.h (hunk 18 above)
Comment (jamesob) [reply src/validation.h:592]: The pointer may actually be nulled out e.g. when we transition the initial IBD chainstate to the background (and give its mempool to the snapshot chainstate, which becomes our active).

Comment (jnewbery) [reply src/validation.h:592]: ah, ok! I'm not totally up to date on how the different chainstate objects work in assumeutxo.

File: src/validation.h (hunk 21 above)
Comment (jamesob) [reply src/validation.h:811]: Good call - weird, thought I had made that change... maybe lost in rebase shuffle.

File: src/validation.cpp (hunk 20 above)
Comment (jamesob) [reply src/validation.cpp]: Fixed, thanks!

File: src/validation.cpp (hunk 17 above)
Comment (jamesob) [reply src/validation.cpp]: Sounds good, fixed.

File: src/validation.cpp (hunk 16 above)
Comment (jamesob) [reply src/validation.cpp]: Good catch, thanks.

**Review by ryanofsky — APPROVED** [review]
Code review ACK c13e832af4a044a5c4beb4694022df53fa2eaf6d

File: src/validation.h [hunk 22]
```diff
@@ -798,6 +800,33 @@ class CChainState
...
+     * tip/height.
+     *
+     * Note: we assume that disconnectpool only contains transactions that are NOT
+     * confirmed in the current chain nor already in the mempool (otherwise,
+     * in-mempool descendants of such transactions would be removed).
+     *
+     * Passing fAddToMempool=false will skip trying to add the transactions back,
+     * and instead just erase from the mempool as needed.
+     */
+    void TryUpdateMempoolForReorg(
```
Comment (jnewbery) [inline src/validation.h:822]: This new name seems worse to me than the original (`UpdateMempoolForReorg()`). The `Try` suggests that this operation may fail or that it's best-effort. In fact, if a mempool exists then it really will be updated to be internally consistent.

File: src/validation.cpp [hunk 23]
```diff
@@ -2218,11 +2213,11 @@ static void UpdateTip(CTxMemPool& mempool, const CBlockIndex* pindexNew, const C
     }
 
     bilingual_str warning_messages;
-    if (!active_chainstate.IsInitialBlockDownload()) {
+    if (!this->IsInitialBlockDownload()) {
```
Comment (jnewbery) [inline src/validation.cpp:2216]: There's no need to use `this->` to call member functions. `this->` is implicitly added before the names: https://en.cppreference.com/w/cpp/language/this.

File: src/validation.cpp [hunk 24]
```diff
@@ -2237,14 +2232,14 @@ static void UpdateTip(CTxMemPool& mempool, const CBlockIndex* pindexNew, const C
       pindexNew->GetBlockHash().ToString(), pindexNew->nHeight, pindexNew->nVersion,
       log(pindexNew->nChainWork.getdouble())/log(2.0), (unsigned long)pindexNew->nChainTx,
       FormatISO8601DateTime(pindexNew->GetBlockTime()),
-      GuessVerificationProgress(chainParams.TxData(), pindexNew), active_chainstate.CoinsTip().DynamicMemoryUsage() * (1.0 / (1<<20)), active_chainstate.CoinsTip().GetCacheSize(),
+      GuessVerificationProgress(m_params.TxData(), pindexNew), this->CoinsTip().DynamicMemoryUsage() * (1.0 / (1<<20)), this->CoinsTip().GetCacheSize(),
```
Comment (jnewbery) [inline src/validation.cpp:2235]: As above, no need for `this->` here.

File: src/validation.h [hunk 25]
```diff
@@ -600,7 +601,10 @@ class CChainState
     //! CChainState instances.
     BlockManager& m_blockman;
 
-    explicit CChainState(CTxMemPool& mempool, BlockManager& blockman, std::optional<uint256> from_snapshot_blockhash = std::nullopt);
+    explicit CChainState(
+        CTxMemPool* const mempool,
```
Comment (jnewbery) [inline src/validation.h:605]: These new `const` keywords on the `CChainstate::CChainstate` and `InitializeChainstate` arguments are harmless but unnecessary. When you're passing by value (which is what happens when you pass a pointer), making that argument const doesn't change the interface, since the argument is copied to a locally scoped parameter inside the function. It's only when passing by reference that marking arguments const matters.

You can tell that it doesn't make a difference because the function declaration here has the first arg as `CTxMemPool* const`, and the function definition in the .cpp file has `CTxMemPool*` (non const pointer). The compiler doesn't complain because the const-ness of a pass by value arg isn't part of the signature.

File: src/validation.cpp [hunk 26]
```diff
@@ -372,12 +365,16 @@ static void UpdateMempoolForReorg(CChainState& active_chainstate, CTxMemPool& me
...
+    m_mempool->UpdateTransactionsFromBlock(vHashUpdate);
 
     // We also need to remove any now-immature transactions
-    mempool.removeForReorg(active_chainstate, STANDARD_LOCKTIME_VERIFY_FLAGS);
+    m_mempool->removeForReorg(*this, STANDARD_LOCKTIME_VERIFY_FLAGS);
     // Re-limit mempool size, in case we added any transactions
-    LimitMempoolSize(mempool, active_chainstate.CoinsTip(), gArgs.GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000, std::chrono::hours{gArgs.GetArg("-mempoolexpiry", DEFAULT_MEMPOOL_EXPIRY)});
+    LimitMempoolSize(
+        *m_mempool,
+        this->CoinsTip(),
```
Comment (jnewbery) [inline src/validation.cpp:375]: No need for `this->`

File: src/test/validation_flush_tests.cpp [hunk 27]
```diff
@@ -20,7 +20,7 @@ BOOST_AUTO_TEST_CASE(getcoinscachesizestate)
 {
     CTxMemPool mempool;
     BlockManager blockman{};
-    CChainState chainstate{mempool, blockman};
+    CChainState chainstate{&mempool, blockman};
```
Comment (jnewbery) [inline src/test/validation_flush_tests.cpp:23]: Any idea why there's a different `tx_pool` from `mempool` in this test? It's not used anywhere so presumably can be removed?

File: src/validation.cpp (hunk 23 above)
Comment (jamesob) [reply src/validation.cpp:2216]: I prefer the `this->` prefix to be explicit about a member function call; last time I talked about this with others it was uncontroversial.

**Comment by jamesob** [comment]
Anyone have any idea what the deal is with libsecp tests failing on ARM here? 
image

This doesn't look like a spurious failure.

File: src/validation.h (hunk 22 above)
Comment (jamesob) [reply src/validation.h:822]: If you felt this way it would have been helpful for you to chime in on @naumenkogs's original comment; doing commits back and forth on relatively cosmetic things like method names is not a good use of energy.

File: src/test/validation_flush_tests.cpp (hunk 27 above)
Comment (jamesob) [reply src/test/validation_flush_tests.cpp:23]: Yep, looks like it. I think when I wrote these I wanted to cover the fact that you could pass in a mempool reference that wasn't owned by the chainstate, but I don't think there's any point in doing that now.

File: src/validation.cpp (hunk 23 above)
Comment (jnewbery) [reply src/validation.cpp:2216]: Curious. I've never come across this style before.

File: src/validation.h (hunk 25 above)
Comment (jnewbery) [reply src/validation.h:605]: > this makes it clear that we aren't going to do any pointer arithmetic on mempool

`mempool` inside the function is a copy of the pointer, and so whether or not it's mutated is not part of the contract with the caller, and so doesn't need to be in the function declaration. Like I said it's harmless, but you'll see that there are almost no examples in the code of pass by value parameters being marked const.

File: src/validation.h (hunk 22 above)
Comment (jnewbery) [reply src/validation.h:822]: Sorry - I didn't see the original comment. Ultimately it's up to you what you put in your PR. I don't think that renaming this function is necessary, and is actually worse for the reasons I gave here, but if other people prefer the new name, that's also fine!

**Comment by ryanofsky** [comment]
@MarcoFalke would probably be most knowledgeable about the qemu-arm coredump https://github.com/bitcoin/bitcoin/pull/22415#issuecomment-879089245 https://cirrus-ci.com/task/6754077301276672?logs=ci#L3494, I was curious about that too. It seems like it is happening in a libsecp256k1 test so not related? But maybe there are ways to debug.

(I don't want to bikeshed the style issues, since they don't actually matter at all. But just for fun I'll say I do agree with John in not liking spurious `const`. Also would drop `Try` prefix or replace it with `Maybe` since `Try` suggests some kind of failure would be expected. I do think having `this->` prefixes is more readable in current code to distinguish method calls from function calls, but in non-legacy code it'd be cleaner to use `lowerCaseMethod()` and `UpperCaseFunction()` names to avoid the ambiguity.)

**Comment by MarcoFalke** [comment]
The `(~x[m][shift]) << (64 - (1 << usebits))) == 0` failure is expected: https://github.com/bitcoin-core/secp256k1/issues/610#issuecomment-482803876

**Comment by MarcoFalke** [comment]
Could edit OP to remove the section that no longer applies to make sure it doesn't end up in the final merge commit?

**Comment by jamesob** [comment]
Pushed a rebase:
- `TryUpdateMempool...` -> `MaybeUpdateMempool...`
- Removed unnecessary `const`s
- Removed outdated content from OP

**Comment by jamesob** [comment]
- unnecessary `tx_pool` removed

**Comment by jnewbery** [comment]
ACK ceb7b35a39145717e2d9d356fd382bd1f95d2a5a

**Review by ryanofsky — APPROVED** [review]
Code review ACK ceb7b35a39145717e2d9d356fd382bd1f95d2a5a (just minor style and test tweaks since last review)

**Review by lsilva01 — APPROVED** [review]
Code review ACK and tested on Signet ACK https://github.com/bitcoin/bitcoin/pull/22415/commits/ceb7b35a39145717e2d9d356fd382bd1f95d2a5a

**Comment by naumenkogs** [comment]
ACK ceb7b35a39145717e2d9d356fd382bd1f95d2a5a

**Review by maflcko — APPROVED** [review]
review ACK ceb7b35a39145717e2d9d356fd382bd1f95d2a5a 😌

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK ceb7b35a39145717e2d9d356fd382bd1f95d2a5a 😌
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhhlwwAtTlbDcsq2EyaUfR0ZvA00PHL6azSeUF1jZ0G0i3eTSVzYlATm+ePtd0W
wsJW4atryMHoqAGZeXTpjt8SZ9XqVdlylswiYefDVTmc9qq9T7ejW4uVtzY/sM45
D6jWRRSUg9xX8c7zrjRaxbPfdGonDTKxkh4BK3v1B9mm2+YVb5EjMDqszzsJs1ef
3VfWNwhRWmfLI9voCAIfOUzFQQC1dWI7NqS7V6MoTnEIpJouK34hvaA51un7iu38
ER5rUhmVUTT/VxkwbzlOlu4C654IVgAq4S2YQ0NWSIDgsuqxEe/C2nfUlm/bPl/v
Khu3jCJz9u26BCLOfkBEpftGUhhFK+djxgitDz/ktzD8taU6Fy2sDPp8vjL8cL3B
//Y87WELvtdUwDxBtqfxkkh+Yy7029uNm3n4Rh6Gex3dsAAZhWAaCKM/1uO1g2+9
KLFtb8sjMb57JQwG3Cdqs889JKjkUzeHdM52w9eu2WBQyWUhSAApuYsf1CUasjwS
I/SNB2Cz
=pjF5
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `74406e84f80bcb1056122ddab2c038117cd73fd1e1a79cd5df1958eacc8b9b20  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010874406e84f80bcb1056122ddab2c038117cd73fd1e1a79cd5df1958eacc8b9b20f0109c605aba88c209bcfaf68ef2cd250c5a08fff0102f80c5ffb12f835de1a9152ae6089ae908f10460f01ce3f008c20260ea6354d9250083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0107325b367d3aafacd81ff8016852979e108f10460f01ce3f008d0ce308234899ed30083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010c8f1833d62954fd0f7e060c589a434d508f10460f01ce3f0083ba32d2949d748bf0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010d94f0bb3e6e0fc83df886f58e081742008f10460f01ce4f00859f0546c2b88d81f0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>
